File: golang/arrays/single_cycle.go
package arrays

func HasSingleCycle(array []int) bool {
	return hasSingleCycle(array, 0)
}

func hasSingleCycle(array []int, startingIndex int) bool {
	// O(n) time | O(1) space
	elementsVisited := 0
	current := startingIndex

	for elementsVisited < len(array) {
		if elementsVisited > 0 && current == startingIndex {
			return false
		}
		elementsVisited++
		current = getNextIdx(current, array)
	}

	return current == startingIndex
}

func getNextIdx(current int, array []int) int {
	jump := array[current]

	if nextIdx := (current + jump) % len(array); nextIdx >= 0 {
		return nextIdx
	} else {
		return nextIdx + len(array)
	}
}

File: python/single_cycle.py
def hasSingleCycle(array):
    visited = 0
    current = 0
    while visited < len(array):
        if visited > 0 and current == 0:
            return False
        visited += 1	
        current = getNextIdx(current, array)
    return current == 0

def getNextIdx(current, array):
    # We want to see how many indcies we have to jump through	
    jump = array[current]
    nextIdx = (current + jump) % len(array)
    # Claify if we can negative integers, can we have decimasls?
     return nextIdx if nextIdx >= 0 else nextIdx + len(array)

